CatAnalyzer/plugins/TtbarDiLeptonAnalyzer.cc

```diff
@@ -270,8 +405,318 @@ bool TtbarDiLeptonAnalyzer::eventSelection(const edm::Event& iEvent, systematic
+        gen_d0s.push_back( ToTLorentzVector(aGenParticle));
+
+        nIDMother = isFromtop(aGenParticle);
+
+        if ( /*0 == 1 &&*/ abs(nIDMother) != 6 ) {
+            continue;
+        }
+
+        //printf("Sign : %i\n", nIDMother * aGenParticle.pdgId() / abs(aGenParticle.pdgId()));
+        //printf("%s\n", ( aGenParticle.charge() * aGenParticle.pdgId() >= 0 ? "S" : "O" ));
+        vecSumMom = ToTLorentzVector(aGenParticle) +
+          ( nIDMother * b_lep1_pid < 0 ? b_lep1 : b_lep2 );
+        b_d0_lepSV_correctM.push_back(vecSumMom.M());
+      } else if ( std::abs(aGenParticle.pdgId()) ==  413 ) {
+        gen_dstars.push_back( ToTLorentzVector(aGenParticle));
+
+        nIDMother = isFromtop(aGenParticle);
+
+        if ( /*0 == 1 &&*/ abs(nIDMother) != 6 ) {
+            continue;
+        }
+
+        //printf("Sign : %i\n", nIDMother * aGenParticle.pdgId() / abs(aGenParticle.pdgId()));
+        //printf("%s\n", ( aGenParticle.charge() * aGenParticle.pdgId() >= 0 ? "S" : "O" ));
+
+        vecSumMom = ToTLorentzVector(aGenParticle) +
+          ( nIDMother * b_lep1_pid < 0 ? b_lep1 : b_lep2 );
+        b_dstar_lepSV_correctM.push_back(vecSumMom.M());
+      } else if ( std::abs(aGenParticle.pdgId()) ==  443 ) {
+        gen_Jpsis.push_back( ToTLorentzVector(aGenParticle));
+
+        nIDMother = isFromtop(aGenParticle);
+
+        if ( /*0 == 1 &&*/ abs(nIDMother) != 6 ) {
+            continue;
+        }
+
+        //printf("Sign : %i\n", nIDMother * aGenParticle.pdgId() / abs(aGenParticle.pdgId()));
+        //printf("%s\n", ( aGenParticle.charge() * aGenParticle.pdgId() >= 0 ? "S" : "O" ));
+
+        vecSumMom = ToTLorentzVector(aGenParticle) +
+          ( nIDMother * b_lep1_pid < 0 ? b_lep1 : b_lep2 );
+        b_Jpsi_lepSV_correctM.push_back(vecSumMom.M());
+      }
+    }
+  }
+  int d0_count=-1;
+  int dstar_count=-1;
+  int Jpsi_count=-1;
+
+  TClonesArray& br_d0 = *b_d0;
+  TClonesArray& br_d0_dau1 = *b_d0_dau1;
+  TClonesArray& br_d0_dau2 = *b_d0_dau2;
+
+  TClonesArray& br_dstar = *b_dstar;
+  TClonesArray& br_dstar_dau1 = *b_dstar_dau1;
+  TClonesArray& br_dstar_dau2 = *b_dstar_dau2;
+  TClonesArray& br_dstar_dau3 = *b_dstar_dau3;
+
+  TClonesArray& br_Jpsi = *b_Jpsi;
+  TClonesArray& br_Jpsi_dau1 = *b_Jpsi_dau1;
+  TClonesArray& br_Jpsi_dau2 = *b_Jpsi_dau2;
+
+  TLorentzVector vecDMMom, vecDau12;
+  float fQDau, fQDM;
+
+  TLorentzVector vecSumDMLep1, vecSumDMLep2;
+  float fMDMLep1, fMDMLep2;
+
+  float fDeltaEta, fDeltaPhi;
+  float fSqrtdRMLep1, fSqrtdRMLep2;
+
+  for( auto& x : *d0s) {
+    d0_count++;
+
+    auto d0_tlv = ToTLorentzVector(x);
+    new( br_d0[d0_count]) TLorentzVector(d0_tlv);
+    new( br_d0_dau1[d0_count]) TLorentzVector(ToTLorentzVector(*(x.daughter(0))));
+    new( br_d0_dau2[d0_count]) TLorentzVector(ToTLorentzVector(*(x.daughter(1))));
+
+    b_d0_dca.push_back( x.dca());
+    double d0_vProb = x.vProb();
+    b_d0_vProb.push_back(d0_vProb);
+
+    if ( abs( d0_vProb ) > 1e-5 ) {
```

Comment:
I know that vProb must be positive values. The abs function was added when vProb function was not stable.